cdc/sinkv2/eventsink/txn/worker.go

```diff
@@ -67,49 +82,91 @@ func (w *worker) runBackgroundLoop() {
+					zap.Int("workerID", w.ID),
 					zap.Error(err))
 			}
 		}()
-		w.timer = time.NewTimer(w.backend.MaxFlushInterval())
+		defer func() {
+			var r interface{}
+			if r = recover(); r == nil {
+				return
+			}
+			buf := make([]byte, 4096)
+			stackSize := runtime.Stack(buf, false)
+			buf = buf[:stackSize]
+			log.Error("Transaction sink worker panics", zap.Reflect("r", r), zap.Stack("stacktrace"))
+			w.errCh <- cerror.ErrMySQLWorkerPanic.GenWithStack("Transaction sink worker panics, stack: %v", string(buf))
+		}()
+		log.Info("Transaction sink worker starts",
+			zap.String("changefeedID", w.changefeed),
+			zap.Int("workerID", w.ID))
+
+		w.timer = time.NewTimer(w.flushInterval)
+		wantMoreCallbacks := make([]func(), 0, 1024)
+	LOOP:
 		for {
 			select {
+			case <-w.ctx.Done():
+				log.Info("Transaction sink worker exits as canceled",
+					zap.String("changefeedID", w.changefeed),
+					zap.Int("workerID", w.ID))
+				return
 			case <-w.stopped:
-				log.Info("Transaction sink backend worker exits expectedly",
+				log.Info("Transaction sink worker exits as closed",
+					zap.String("changefeedID", w.changefeed),
 					zap.Int("workerID", w.ID))
 				return
 			case txn := <-w.txnCh.Out():
 				metrics.ConflictDetectDuration.Observe(time.Since(txn.start).Seconds())
-				txn.wantMore()
-				if w.backend.OnTxnEvent(txn.txnEvent.TxnCallbackableEvent) && w.doFlush() {
-					log.Warn("Transaction sink backend exits unexceptedly")
-					return
+				wantMoreCallbacks = append(wantMoreCallbacks, txn.wantMore)
+				if w.backend.OnTxnEvent(txn.txnEvent.TxnCallbackableEvent) {
+					if w.doFlush() {
+						break LOOP
+					}
+				} else {
+					continue LOOP
 				}
 			case <-w.timer.C:
 				if w.doFlush() {
-					log.Warn("Transaction sink backend exits unexceptedly")
-					return
+					break LOOP
 				}
 			}
+			// Flush successfully, call callbacks to notify conflict detector.
+			for _, wantMore := range wantMoreCallbacks {
+				wantMore()
+			}
+			wantMoreCallbacks = wantMoreCallbacks[:0]
+			if cap(wantMoreCallbacks) > 1024 {
+				wantMoreCallbacks = make([]func(), 0, 1024)
+			}
 		}
+		log.Warn("Transaction sink worker exits unexceptedly",
+			zap.String("changefeedID", w.changefeed),
+			zap.Int("workerID", w.ID))
 	}()
 }
 
 // doFlush flushes the backend. Returns true if the goroutine can exit.
 func (w *worker) doFlush() bool {
-	// TODO: support to cancel the worker when performing some blocking operations.
-	ctx := context.Background()
-	if err := w.backend.Flush(ctx); err != nil {
-		log.Warn("txn sink worker flush fail", zap.Error(err))
+	if err := w.backend.Flush(w.ctx); err != nil {
+		log.Warn("Transaction sink backend flush fail",
+			zap.String("changefeedID", w.changefeed),
+			zap.Int("workerID", w.ID),
+			zap.Error(err))
 		select {
+		case <-w.ctx.Done():
 		case w.errCh <- err:
-		case <-ctx.Done():
 		}
 		return true
 	}
 	if !w.timer.Stop() {
-		<-w.timer.C
+		select {
+		case <-w.timer.C:
+		default:
```

Comment:
Why do we need a default branch here?

Reply:
Yes. Other wise it will be blocked if the timer has already been triggered.